## Motivation  

The **ECO Platform API** provides developers with seamless access to high-quality, standardized Environmental Product Declaration (EPD) data. As the construction sector moves towards **low-carbon solutions and resource efficiency**, reliable and scientifically validated data is essential for **LCA modeling, sustainable decision-making, and regulatory compliance**.  

ECO Platform brings together EPD program operators, manufacturers, and sustainability experts to ensure **harmonized, credible, and transparent environmental data**. This API is designed to integrate effortlessly into your **LCA workflows**, enabling automation, data-driven analysis, and innovative sustainability applications.  

By leveraging the ECO Platform API, you can:  
- **Access standardized EPD data** for accurate environmental impact assessments.  
- **Automate LCA calculations** and integrate with existing sustainability tools.  
- **Ensure compliance** with international sustainability standards.  
- **Enhance transparency** in construction and green building projects.  

This documentation will guide you through **authenticating, retrieving, and processing EPD data**, providing practical examples to help you integrate the API into your workflows efficiently.  

### Target Audience  

This documentation is for developers and professionals **familiar with Python, APIs, and well-versed in LCA and EPD data**.  

We also assume that you:  
✅ **Have Python set up** ([Setup guide](#python-setup))  
✅ **Have an ECO Platform API key** ([Get your key here](#getting-api-key))

In [1]:
!pip install pandoc requests dotenv

In [2]:
import requests
import pandas as pd

### Load the Token

In [3]:
# past you token here
TOKEN = "CHANGE ME TO THE TOKEN"

# or load form env
if TOKEN.startswith("CHANGE"):
    import os
    from dotenv import load_dotenv

    load_dotenv()
    # load the Access Token
    TOKEN = os.getenv('TOKEN')

# Verify token is set
if not TOKEN:
    raise Exception("Error: TOKEN is not set. Please set the TOKEN variable.")

## Get Global Warming Potential (GWP) of Concrete

### Step 1: Find the Concrete 

In [4]:

# Base URL for the Eco Platform data aggregator
BASE_URL = 'https://data.eco-platform.org/resource/'

# API endpoint for fetching process data
PROCESS_ENDPOINT = 'processes'

# Query parameters for filtering concrete data
query_params = {
    'search': 'true',
    'distributed': 'true',
    'virtual': 'true',
    'metaDataOnly': 'false',
    'validUntil': '2024',
    'pageSize': 5,  
    'startIndex': 0,
    'format': 'JSON',
}
headers = {
    'Authorization': f'Bearer {TOKEN}'  
}

# Specify the name of the concrete mix to search for
query_params["name"] = "Concrete C25/30 XC4 XF1 XA1 F3 16 M ECOPact , Mix number DA4234-DHFS Version 1"

# Send GET request to fetch relevant concrete data
response = requests.get(f'{BASE_URL}{PROCESS_ENDPOINT}', headers=headers, params=query_params)
response.raise_for_status()

# Convert response data into a structured DataFrame
concrete_data_df = pd.DataFrame(response.json()["data"])
concrete_data_df

languages                                               name  \
0  [de, en]  Concrete C25/30 XC4 XF1 XA1 F3 16 L ECOPact, M...   
1  [de, en]  Concrete C25/30 XC4 XF1 XA1 F3 16 L ECOPactR, ...   
2  [de, en]  Concrete C25/30 XC4 XF1 XA1 F3 16 M ECOPact , ...   
3  [de, en]  Concrete C25/30 XC4 XF1 XA1 F3 16 M ECOPact , ...   
4  [de, en]  Concrete C25/30 XC4 XF1 XA1 F3 16 M ECOPact , ...   

                                   uuid    version geo  \
0  0c407bf2-c1f2-4a26-ace4-9c0fb1117e3f  00.01.001  DE   
1  24484f95-db48-4d40-9b06-c08ab7b4cd51  00.01.001  DE   
2  02e4fdcc-84d7-4f19-95e6-e8d19de9d6ec  00.01.001  DE   
3  0ebf7111-4d1d-47a9-96cd-2e969f0f96a3  00.01.001  DE   
4  037f6652-34a2-4019-a360-535a6334313a  00.01.001  DE   

                                           classific classificSystem type  \
0  02 Bauprodukte / Normal- Leicht- und Porenbeto...   IBUCategories  EPD   
1  02 Bauprodukte / Normal- Leicht- und Porenbeto...   IBUCategories  EPD   
2  02 Bauprodukte / Normal- Leicht- und Porenbeto...   IBUCategories  EPD   
3  02 Bauprodukte / Normal- Leicht- und Porenbeto...   IBUCategories  EPD   
4  02 Bauprodukte / Normal- Leicht- und Porenbeto...   IBUCategories  EPD   

   refYear  validUntil                                         compliance  \
0     2023        2028  [{'name': 'EN 15804+A2', 'uuid': 'c0016b33-8cf...   
1     2023        2028  [{'name': 'EN 15804+A2', 'uuid': 'c0016b33-8cf...   
2     2023        2028  [{'name': 'EN 15804+A2', 'uuid': 'c0016b33-8cf...   
3     2023        2028  [{'name': 'EN 15804+A2', 'uuid': 'c0016b33-8cf...   
4     2023        2028  [{'name': 'EN 15804+A2', 'uuid': 'c0016b33-8cf...   

            subType                                        dataSources  \
0  specific dataset  [{'uuid': '28d74cc0-db8b-4d7e-bc44-5f6d56ce0c4...   
1  specific dataset  [{'uuid': '28d74cc0-db8b-4d7e-bc44-5f6d56ce0c4...   
2  specific dataset  [{'uuid': '28d74cc0-db8b-4d7e-bc44-5f6d56ce0c4...   
3  specific dataset  [{'uuid': '28d74cc0-db8b-4d7e-bc44-5f6d56ce0c4...   
4  specific dataset  [{'uuid': '28d74cc0-db8b-4d7e-bc44-5f6d56ce0c4...   

                                         regNo  \
0  IBU-HOL-HOL-2204100-DE2023000014-1SUG001-DE   
1  IBU-HOL-HOL-2204100-DE2023000026-1SUG001-DE   
2  IBU-HOL-HOL-2204100-DE2023000046-1SUG001-DE   
3  IBU-HOL-HOL-2204100-DE2023000034-1SUG001-DE   
4  IBU-HOL-HOL-2204100-DE2023000039-1SUG001-DE   

                                       regAuthority    nodeid  \
0  {'uuid': 'd111dbec-b024-4be5-86c5-752d6eb2cf95'}  IBU_DATA   
1  {'uuid': 'd111dbec-b024-4be5-86c5-752d6eb2cf95'}  IBU_DATA   
2  {'uuid': 'd111dbec-b024-4be5-86c5-752d6eb2cf95'}  IBU_DATA   
3  {'uuid': 'd111dbec-b024-4be5-86c5-752d6eb2cf95'}  IBU_DATA   
4  {'uuid': 'd111dbec-b024-4be5-86c5-752d6eb2cf95'}  IBU_DATA   

                                                 uri   dsType  
0  https://ibudata.lca-data.com/resource/datastoc...  Process  
1  https://ibudata.lca-data.com/resource/datastoc...  Process  
2  https://ibudata.lca-data.com/resource/datastoc...  Process  
3  https://ibudata.lca-data.com/resource/datastoc...  Process  
4  https://ibudata.lca-data.com/resource/datastoc...  Process

### Step 2: Retrieve detailed Environmental Product Declaration (EPD) data

In [5]:
# URL to fetch detailed information on EPD
EPD_URL = "https://ibudata.lca-data.com/resource/processes/"

detail_params = {
    "format": "JSON",
    "view": "extended"
}

# Unique identifier for the selected concrete data entry
concrete_uuid = "0ebf7111-4d1d-47a9-96cd-2e969f0f96a3"

# Fetch extended details of the selected concrete mix
response = requests.get(f'{EPD_URL}{concrete_uuid}', headers=headers, params=detail_params)
response.raise_for_status()

detailed_data = response.json()

# Display available data keys for exploration
list(detailed_data.keys())

['processInformation',
 'modellingAndValidation',
 'administrativeInformation',
 'exchanges',
 'LCIAResults',
 'otherAttributes',
 'version']

### Step 3: Extract and Print GWP

In [6]:
# Access the first LCIA (Life Cycle Impact Assessment) result, which represents GWP-Total
gwp_result = detailed_data["LCIAResults"]["LCIAResult"][0]

# Extract short description of the impact category (e.g., "GWP-total")
gwp_description = gwp_result["referenceToLCIAMethodDataSet"]["shortDescription"][0]["value"]
print(gwp_description)

# Extract unit of measurement for GWP (NOTE index 0 is always the description)
gwp_units = gwp_result["other"]["anies"]
unit_description = gwp_units[0]["value"]["shortDescription"][0]["value"]


# Calculate the total GWP value
total_gwp_value = sum(float(entry["value"]) for entry in gwp_units[1:])
print(f"\nTotal GWP: {total_gwp_value:.1f} {unit_description}")


Global Warming Potential - total (GWP-total)

Total GWP: 166.2 kg CO_(2) eqv.


### Step 4: Retrieve Reference Flow Properties

In [7]:
# Find reference flow index
reference_flow_idx = detailed_data["processInformation"]["quantitativeReference"]["referenceToReferenceFlow"][0]

# Extract relevant flow properties
flow_properties = detailed_data["exchanges"]["exchange"][reference_flow_idx]["flowProperties"]

# Fetch flow properties lookup table
flow_properties_url = "https://data.eco-platform.org/resource/flowproperties"
flow_properties_response = requests.get(flow_properties_url, headers=headers, params={'format': 'JSON'})
flow_properties_lookup = pd.DataFrame(flow_properties_response.json()["data"])

# Process and display flow properties
processed_flow_properties = []
for prop in flow_properties:
    # Look up the property details using UUID
    lookup_result = flow_properties_lookup.loc[flow_properties_lookup["uuid"] == prop["uuid"]]
    
    # Convert lookup result to dictionary format
    lookup_data = {key: list(value.values())[0] for key, value in lookup_result.to_dict().items()}
    print(lookup_data["name"], prop["meanValue"], lookup_data["defUnit"])
    
    processed_flow_properties.append({
        "name": lookup_data["name"],
        "value": prop["meanValue"],
        "unit": lookup_data["defUnit"]
    })

print("\nHas a GWP of:")
print(f"{total_gwp_value:.1f} {unit_description}")

Mass 2310.0 kg
Volume 1.0 m3

Has a GWP of:
166.2 kg CO_(2) eqv.


### Step 5: Calculate GWP for 1 Ton of Concrete

In [8]:
# Extract mass property for scaling calculations
mass_property = next(fp for fp in processed_flow_properties if fp["name"] == "Mass")
conversion_factor = mass_property["value"] / 1000  # Convert from kg to tons

# Compute the total GWP for 1 ton of concrete
print(f"That means 1 ton (1000 kg) of concrete has a GWP of {(total_gwp_value / conversion_factor):.1f} {unit_description}")


That means 1 ton (1000 kg) of concrete has a GWP of 71.9 kg CO_(2) eqv.


### List all Processes

In [9]:
import json
import requests
import pandas as pd


# Base URL of the Aggregator 
base_url = 'https://data.eco-platform.org/resource/'

# Endpoint for processes
endpoint = 'processes'

# Query parameters
params = {
    'search': 'true',
    'distributed': 'true',
    'virtual': 'true',
    'metaDataOnly': 'false',
    'validUntil': '2024',
    'pageSize': 5,
    'startIndex': 0,
    'format': 'JSON',
}
headers = {
    'Authorization': f'Bearer {TOKEN}'
}

response = requests.get(f'{base_url}{endpoint}', headers=headers, params=params)
response.raise_for_status() 

df = pd.DataFrame(response.json()["data"])
df

languages                                               name  \
0      [en]                             Birsta S N2 W2 c/c 1 m   
1      [en]                            Durasein Solid Surface    
2      [no]                                                NaN   
3      [no]                                                NaN   
4      [en]   ABT Skanska Grön AsfaltBio Zero. Gällivare, L...   

                                   uuid    version  geo  \
0  18f4602a-6d0d-46ff-8701-9402b8e70e6f  00.05.001   CN   
1  e26d3411-373a-4cdd-af82-08db7e36184f  07.01.052  GLO   
2  ac58a83c-e898-4505-9d93-9c5b4e3e852f  00.04.000   NO   
3  10eba8d0-fcd6-4c80-9227-69de52a68740  00.03.000   IT   
4  58871b96-bcdf-4438-b82f-d100c1df1fe5  00.03.002   SE   

                                           classific  \
0             Bygg / Stål-, aluminiumskonstruksjoner   
1  Construction products, Infrastructure & buildings   
2                                 Bygg / Betongvarer   
3             Bygg / Stål-, aluminiumskonstruksjoner   
4                              Bygg / Asfalt og pukk   

                            classificId               classificSystem type  \
0  992467c5-eec4-4d69-88ba-672f1a896481                      EPDNorge  EPD   
1                                     0  The International EPD System  EPD   
2  f107d200-244c-4c4b-9a14-0c2264316bc2                      EPDNorge  EPD   
3  992467c5-eec4-4d69-88ba-672f1a896481                      EPDNorge  EPD   
4  67dd3145-6d07-4939-8373-297d512803ca                      EPDNorge  EPD   

   refYear  validUntil                                         compliance  \
0     2024        2029  [{'name': 'EN 15804+A2', 'uuid': 'c0016b33-8cf...   
1     2023        2028  [{'name': 'EN 15804+A2 (EF 3.0)', 'uuid': 'c00...   
2     2024        2029  [{'name': 'EN 15804+A2', 'uuid': 'c0016b33-8cf...   
3     2025        2030  [{'name': 'EN 15804+A2', 'uuid': 'c0016b33-8cf...   
4     2021        2026  [{'name': 'EN 15804', 'uuid': 'b00f9ec0-7874-1...   

            subType                                        dataSources  \
0  specific dataset  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   
1  specific dataset  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   
2  specific dataset  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   
3   generic dataset  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   
4  specific dataset  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   

                             regNo  \
0                NEPD-6941-6333-EN   
1  EPD-IES-0007415:004 (S-P-07415)   
2                NEPD-8487-8141-NO   
3                   NEPD-8650-8317   
4                   NEPD-2820-1517   

                                       regAuthority           nodeid  \
0  {'uuid': '7008caa3-6fea-4ad2-81eb-f2b7a0610e6c'}  EPD-NORWAY_DIGI   
1  {'uuid': '5e09bd06-11a1-4292-a56b-47898ad6a691'}       ENVIRONDEC   
2  {'uuid': '7008caa3-6fea-4ad2-81eb-f2b7a0610e6c'}  EPD-NORWAY_DIGI   
3  {'uuid': '7008caa3-6fea-4ad2-81eb-f2b7a0610e6c'}  EPD-NORWAY_DIGI   
4  {'uuid': '7008caa3-6fea-4ad2-81eb-f2b7a0610e6c'}  EPD-NORWAY_DIGI   

                                                 uri   dsType  
0  https://epdnorway.lca-data.com/resource/datast...  Process  
1  https://data.environdec.com/resource/datastock...  Process  
2  https://epdnorway.lca-data.com/resource/datast...  Process  
3  https://epdnorway.lca-data.com/resource/datast...  Process  
4  https://epdnorway.lca-data.com/resource/datast...  Process

In the example above, we sent our first request to the ECO Platform API and got a list of five processes back. We then loaded this response data into a Pandas DataFrame and showed the table.

You can change the parameters, try different endpoints, or further process the data using Pandas. For more details, check the full [API documentation](https://bitbucket.org/okusche/soda4lca/src/7.x-branch/Doc/src/Service_API/Service_API_Datasets_GET.md).

### Filter by **compliance**

[This](https://bitbucket.org/okusche/soda4lca/src/26955bf0d6dabc008e19a39c975973f762ef281f/Doc/src/Service_API/Service_API_Datasets_Process_Query.md?at=7.x-branch#Service_API_Datasets_Process_Query.md-44) is the API documentation for it 

The UUIDs for compliance are from [here](https://indatawg.github.io/ILCD-EPD-Data-Format/gitBranches/main/identifiers/Common_references.html) or below

|Name| UUID|
|----|-----|	
|ILCD Format	|	a97a0155-0234-4b87-b4ce-a45da52f2a40	|
|EPD Data Format Extensions v1.1	|	cba73800-7874-11e3-981f-0800200c9a66	|
|EPD Data Format Extensions v1.2	|	a29449fd-aa2f-4de8-b5d7-4b06b43c6fde	|
|EN 15804+A1	|	b00f9ec0-7874-11e3-981f-0800200c9a66	|
|EN 15804+A2 (EF 3.0)	|	c0016b33-8cf7-415c-ac6e-deba0d21440d	|
|EN 15804+A2 (EF 3.1)	|	d4aa3ec7-b1d7-4a4a-a6cb-37af88dcc902	|
|ISO 14025	|	4f2eb655-6e44-4874-a95a-e28f5442cd4d	|
|ISO 21930	|	9c1e8451-2321-4498-b459-b6568b6b6547	|

In [10]:
params["complianceMode"] = "AND"

params["compliance"] = ["b00f9ec0-7874-11e3-981f-0800200c9a66",  "c0016b33-8cf7-415c-ac6e-deba0d21440d"]	
res = requests.get(f'{base_url}{endpoint}', headers=headers, params=params)
res.raise_for_status() 

del params['complianceMode']
del params["compliance"]

df = pd.DataFrame(res.json()["data"])
df

languages                                               name  \
0      [en]   ABT Skanska Grön AsfaltBio Zero. Gällivare, L...   
1      [en]                      Leca® Letklinker 10-20; Hinge   
2      [no]                                                NaN   
3      [en]                     weberbase 130 lättbetongputs B   
4      [no]                                                NaN   

                                   uuid    version geo  \
0  58871b96-bcdf-4438-b82f-d100c1df1fe5  00.03.002  SE   
1  06f4050b-cb0c-4920-978f-9d588f8ce12c  00.13.001  DK   
2  214b178e-8740-47e8-bc3b-515daeaecda3  00.03.000  NO   
3  627b52c5-aee6-443e-84ea-e81a8c825d1a  00.03.001  SE   
4  4895e95e-976e-44b9-a405-36ab28061749  00.04.001  NO   

                                    classific  \
0                       Bygg / Asfalt og pukk   
1                            Bygg / Isolasjon   
2  Bygg / Teknisk-kjemiske byggevareprodukter   
3  Bygg / Teknisk-kjemiske byggevareprodukter   
4                        Bygg / Ferdig betong   

                            classificId classificSystem type  refYear  \
0  67dd3145-6d07-4939-8373-297d512803ca        EPDNorge  EPD     2021   
1  daa889fc-3d34-43d1-b6f6-a3927c8911cd        EPDNorge  EPD     2020   
2  071f9a38-08af-4ee5-909a-9884e93816c0        EPDNorge  EPD     2023   
3  071f9a38-08af-4ee5-909a-9884e93816c0        EPDNorge  EPD     2022   
4  6cf99d9f-7e16-4747-83c3-6e86a60b2ff3        EPDNorge  EPD     2020   

   validUntil                                         compliance  \
0        2026  [{'name': 'EN 15804', 'uuid': 'b00f9ec0-7874-1...   
1        2025  [{'name': 'EN 15804', 'uuid': 'b00f9ec0-7874-1...   
2        2028  [{'name': 'EN 15804', 'uuid': 'b00f9ec0-7874-1...   
3        2027  [{'name': 'EN 15804', 'uuid': 'b00f9ec0-7874-1...   
4        2025  [{'name': 'EN 15804', 'uuid': 'b00f9ec0-7874-1...   

            subType                                        dataSources  \
0  specific dataset  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   
1  specific dataset  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   
2  specific dataset  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   
3  specific dataset  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   
4  specific dataset  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   

               regNo                                      regAuthority  \
0     NEPD-2820-1517  {'uuid': '7008caa3-6fea-4ad2-81eb-f2b7a0610e6c'}   
1     NEPD-2249-1030  {'uuid': '7008caa3-6fea-4ad2-81eb-f2b7a0610e6c'}   
2  NEPD-4268-3502-NO  {'uuid': '7008caa3-6fea-4ad2-81eb-f2b7a0610e6c'}   
3  NEPD-3620-2566-EN  {'uuid': '7008caa3-6fea-4ad2-81eb-f2b7a0610e6c'}   
4  NEPD-2507-1249-NO  {'uuid': '7008caa3-6fea-4ad2-81eb-f2b7a0610e6c'}   

            nodeid                                                uri   dsType  
0  EPD-NORWAY_DIGI  https://epdnorway.lca-data.com/resource/datast...  Process  
1  EPD-NORWAY_DIGI  https://epdnorway.lca-data.com/resource/datast...  Process  
2  EPD-NORWAY_DIGI  https://epdnorway.lca-data.com/resource/datast...  Process  
3  EPD-NORWAY_DIGI  https://epdnorway.lca-data.com/resource/datast...  Process  
4  EPD-NORWAY_DIGI  https://epdnorway.lca-data.com/resource/datast...  Process

In this example, we only chose EPDs that comply with `EN15804+A1` and `+A2 EF 3.0`.

### Select a specific EPD-Number


We use 2 types of Background Databases 


#### [GaBi](https://indatawg.github.io/ILCD-EPD-Data-Format/gitBranches/main/identifiers/BackgroundDB_SourceDatasets_GaBi.html) 

|GaBi Database Version | Name| UUID |
|-----|-------------------------|----------------------------------------|
| all | GaBi Database (general) | 28d74cc0-db8b-4d7e-bc44-5f6d56ce0c4a |
| SP20 2013 | GaBi database SP20 2013 | b9fdf525-e9b8-47a0-b897-2ecfe7e2cff5 |
| SP25 2014 | GaBi database SP25 2014 | 04536297-8f33-4ef0-b06c-54dc273c2aaa |
| SP26 2014 | GaBi database SP26 2014 | 7d81b748-d7ea-4efd-b72e-205c69306f38 |
| SP27 2015 | GaBi database SP27 2015 | 9422129a-7992-4eaf-aa17-0fca36af4512 |
| SP28 2015 | GaBi database SP28 2015 | 176effc6-2e73-4419-be24-60b3399e91cf |
| SP29 2016 | GaBi database SP29 2016 | 8cb50bdc-0299-4a4c-9a4e-4c62917018c7 |
| SP30 2016 | GaBi database SP30 2016 | 6985fea4-15ea-4b96-9237-c67ee3a1f28c |
| SP33 2017 | GaBi database SP33 2017 | d8181616-4417-4dbd-9614-09937353f17a |
| SP34 2017 | GaBi database SP34 2017 | 07650073-f65e-4d07-97e9-359a58b7299e |
| SP35 2018 | GaBi database SP35 2018 | feda7597-12a1-4bd5-93ef-4d8120877b78 |
| SP36 2018 | GaBi database SP36 2018 | dad5023c-8d97-4656-b82a-2fc6480b33a7 |
| SP37 2019 | GaBi database SP37 2019 | aaef5d83-7fd3-424f-86c6-00f51cd39845 |
| SP38 2019 | GaBi database SP38 2019 | 4467e22f-78e4-4e21-ae4a-b8ffeda7d795 |
| SP39 2019 | GaBi database SP39 2019 | 07d9b8a7-e0fc-408f-a95e-c90735f843ce |
| SP40 2020 | GaBi database SP40 2020 | 0115289e-52f0-4920-bc7e-57b1c96aec8d |
| Edition 2020, CUP 2020.02 | GaBi Database Edition 2020, CUP 2020.02 | 4b368ada-d799-4afc-9d41-f371f2b23781 |
| Edition 2021, CUP 2021.01 | GaBi Database Edition 2021, CUP 2021.01 | 1cece811-7501-44d4-8067-9259bd5cb835 |
| Edition 2021, CUP 2021.02 | GaBi Database Edition 2021, CUP 2021.02 | 90321ffe-9ebb-4946-886c-338efe6fbf26 |
| Edition 2022, CUP 2022.01 | GaBi Database Edition 2022, CUP 2022.01 | 86f664e2-7dab-4722-aad7-4c5206a4123b |
| Edition 2022, CUP 2022.02 | GaBi Database Edition 2022, CUP 2022.02 | 94f442cb-662e-43ea-a792-4b0a35132e93 |
| CUP 2023.01	 | Sphera MLC (fka GaBi) CUP 2023.01 | 6bcd5bfa-3623-48ab-a52b-6d5be8db82ed |
| CUP 2023.02	 | Sphera MLC (fka GaBi) CUP 2023.02 | cb600d3b-e85a-42b3-b059-5591e2757fbb |
| Edition 2024 | Sphera MLC (fka GaBi) Databases Edition 2024 | 7443b962-0446-4d69-8aa7-8055a6b1614a |


#### and [ecoinvent](https://indatawg.github.io/ILCD-EPD-Data-Format/gitBranches/main/identifiers/BackgroundDB_SourceDatasets_ecoinvent.html)


|ecoinvent Database Version | Name | UUID |
|---------------------------|------|------|
|all | ecoinvent database (general) | b497a91f-e14b-4b69-8f28-f50eb1576766 |
|2.0 | ecoinvent 2.0 database | 67e57f6b-53d2-4adb-a5db-7e8157ce5865 |
|2.1 | ecoinvent 2.1 database | dd4ba261-1847-4d00-a9c0-7e81a3402dae |
|2.2 | ecoinvent 2.2 database | 9ffadc15-4649-4871-b565-cee8197b17e9 |
|3.0 | ecoinvent 3.0 database | ae9fac17-61c1-4776-947a-0fc2865954ff |
|3.1 | ecoinvent 3.1 database | 44df4cc3-a75f-4e13-a37a-971dadc97b35 |
|3.2 | ecoinvent 3.2 database | d82a6cdc-1930-4a8a-9b07-2f7fda03fcbd |
|3.3 | ecoinvent 3.3 database | c667cc42-12a7-4d92-9d7b-5e4e90a7a274 |
|3.4 | ecoinvent 3.4 database | db59e0c6-db9a-4d9a-8637-67b8ed42415c |
|3.5 | ecoinvent 3.5 database | ca0da4f4-c837-4005-9c17-e77b9bd09d8b |
|3.6 | ecoinvent 3.6 database | 766b9973-5ee8-48a2-8098-72c8e20b369f |
|3.7 | ecoinvent 3.7 database | de2729d9-235e-4588-8dd0-dba6a1d1f2e1 |
|3.7.1 | ecoinvent 3.7.1 database | 3a4d6cdd-46ad-4ff3-8452-a0323e94f789 |
|3.8 | ecoinvent 3.8 database | 0f4c65c0-6210-450f-b886-82bf8dd5ace9 |
|3.9 | ecoinvent 3.9 database | c6825ce7-2951-4e1e-8634-426c7af4679b |
|3.9.1 | ecoinvent 3.9.1 database | e2034c8f-9d74-46f2-8fce-1f0c5c5cebeb |
|3.10 | ecoinvent 3.10 database | 6edab576-5247-472f-a2f4-0cb84561ca8b |

[This](https://bitbucket.org/okusche/soda4lca/src/26955bf0d6dabc008e19a39c975973f762ef281f/Doc/src/Service_API/Service_API_Datasets_Process_Query.md?at=7.x-branch#Service_API_Datasets_Process_Query.md-44) is the API documentation for it 

In [11]:
params["registrationNumber"] = "EPD-IES"

res = requests.get(f'{base_url}{endpoint}', headers=headers, params=params)
res.raise_for_status()
del params['registrationNumber']

df = pd.DataFrame(res.json()["data"])
df

languages                                               name  \
0      [en]                            Durasein Solid Surface    
1      [en]                           Blended Hydraulic Cement   
2      [en]   CEM II/B-LL 32,5 R - Sacchi - Colacem S.p.A. ...   
3      [en]   CEM IV/A (P) 42,5 R - SR - RAGUSA CEMENTI S.p...   
4      [en]   CEM IV/B (P) 32,5 R - LH/SR - Sfuso - Colacem...   

                                   uuid    version  geo  \
0  e26d3411-373a-4cdd-af82-08db7e36184f  07.01.052  GLO   
1  efe3d139-d74f-46e2-37b4-08dc2e3a4f3f  07.01.009   SA   
2  23832079-c797-4932-456e-08dc6fc9d4ca  07.01.026  GLO   
3  31ae4037-e9e6-48fa-9b9c-08dbdfa8f0c9  07.01.008  GLO   
4  a757fac6-b025-4b0c-c86a-08db8d2354cc  07.01.004  GLO   

                                           classific classificId  \
0  Construction products, Infrastructure & buildings           0   
1                              Construction products           0   
2                              Construction products           0   
3                              Construction products           0   
4  Construction products, Infrastructure & buildings           0   

                classificSystem type  refYear  validUntil  \
0  The International EPD System  EPD     2023        2028   
1  The International EPD System  EPD     2024        2029   
2  The International EPD System  EPD     2024        2029   
3  The International EPD System  EPD     2023        2028   
4  The International EPD System  EPD     2023        2028   

                                          compliance           subType  \
0  [{'name': 'EN 15804+A2 (EF 3.0)', 'uuid': 'c00...  specific dataset   
1  [{'name': 'EN 15804+A2 (EF 3.0)', 'uuid': 'c00...  specific dataset   
2  [{'name': 'EN 15804+A2 (EF 3.0)', 'uuid': 'c00...  specific dataset   
3  [{'name': 'EN 15804+A2 (EF 3.0)', 'uuid': 'c00...  specific dataset   
4  [{'name': 'EN 15804+A2 (EF 3.0)', 'uuid': 'c00...  specific dataset   

                                         dataSources  \
0  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   
1  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   
2  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   
3  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   
4  [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...   

                             regNo  \
0  EPD-IES-0007415:004 (S-P-07415)   
1  EPD-IES-0012616:001 (S-P-12616)   
2  EPD-IES-0013856:001 (S-P-13856)   
3  EPD-IES-0011485:001 (S-P-11485)   
4  EPD-IES-0010140:002 (S-P-10140)   

                                       regAuthority      nodeid  \
0  {'uuid': '5e09bd06-11a1-4292-a56b-47898ad6a691'}  ENVIRONDEC   
1  {'uuid': '5e09bd06-11a1-4292-a56b-47898ad6a691'}  ENVIRONDEC   
2  {'uuid': '5e09bd06-11a1-4292-a56b-47898ad6a691'}  ENVIRONDEC   
3  {'uuid': '5e09bd06-11a1-4292-a56b-47898ad6a691'}  ENVIRONDEC   
4  {'uuid': '5e09bd06-11a1-4292-a56b-47898ad6a691'}  ENVIRONDEC   

                                                 uri   dsType  
0  https://data.environdec.com/resource/datastock...  Process  
1  https://data.environdec.com/resource/datastock...  Process  
2  https://data.environdec.com/resource/datastock...  Process  
3  https://data.environdec.com/resource/datastock...  Process  
4  https://data.environdec.com/resource/datastock...  Process

## Explain the architecture

The ECO Platform has two main parts. The first part collects and combines data and is directly ownd by the ECO Platform. 

The second part consists of the nodes of the EPD Program Operator, which are also connected to the eco platform. 

So far, we have only worked on the Aggregator node.

This diagram explains the relationship.

![Aggregator](./img/ECO_EPD_Portal.004.png)

The Aggregator node does not store data; it only collects data from other nodes and can perform queries on it. It also does not show the complete data. 

If you want to perform more complex queries on a single EPD, you can talk to the node directly to get more detailed information like so.

![full](./img/ECO_EPD_Portal.005.png)


so lets see what we can do on a single node 

### Getting the PDF

In [12]:
lcd = df.iloc[0]
print(lcd)
uri = lcd["uri"]
uuid = lcd["uuid"]
version = lcd["version"]

languages                                                       [en]
name                                         Durasein Solid Surface 
uuid                            e26d3411-373a-4cdd-af82-08db7e36184f
version                                                    07.01.052
geo                                                              GLO
classific          Construction products, Infrastructure & buildings
classificId                                                        0
classificSystem                         The International EPD System
type                                                             EPD
refYear                                                         2023
validUntil                                                      2028
compliance         [{'name': 'EN 15804+A2 (EF 3.0)', 'uuid': 'c00...
subType                                             specific dataset
dataSources        [{'uuid': 'b497a91f-e14b-4b69-8f28-f50eb157676...
regNo                             

In [13]:
print(uri)

# Why??

base_url = uri.split("/")[0] +"//" + uri.split("/")[2] 

url = f"{base_url}/resource/processes/{uuid}/epd?version={version}"
print(url)

# Make the GET request
response = requests.get(url, headers=headers)

# Check if the request was successful
if response.status_code == 200:
    # Save the PDF to a file
    with open("example_process.pdf", "wb") as file:
        file.write(response.content)
    print("PDF downloaded successfully.")
else:
    print(f"Failed to download PDF. Status code: {response.status_code}")


https://data.environdec.com/resource/datastocks/a6c533b3-502e-47b9-885d-31304bf15c64/processes/e26d3411-373a-4cdd-af82-08db7e36184f?version=07.01.052
https://data.environdec.com/resource/processes/e26d3411-373a-4cdd-af82-08db7e36184f/epd?version=07.01.052
PDF downloaded successfully.


### Get CSV

In [14]:
# export https://bitbucket.org/okusche/soda4lca/src/0c04c255287f007249929e9fa80a1ea19f60601b/Doc/src/Service_API/Service_API_Datastock_GET_Export.md
base_url = uri.split("/")[0] +"//" + uri.split("/")[2] 
base_url = "https://data.environdec.com/"
url = f"{base_url}/resource/datastocks/{uuid}/exportCSV"
print(url)

# Make the GET request
response = requests.get(url , headers=headers)

print(response.text)


# Check if the request was successful
# if response.status_code == 200:
#     # Save the CSV to a file
#     with open("example_process.csv", "wb") as file:
#         file.write(response.content)
#     print("CSV downloaded successfully.")
# else:
#     print(f"Failed to download CSV. Status code: {response.status_code}")

https://data.environdec.com//resource/datastocks/e26d3411-373a-4cdd-af82-08db7e36184f/exportCSV
<!doctype html><html lang="en"><head><title>HTTP Status 501 – Not Implemented</title><style type="text/css">body {font-family:Tahoma,Arial,sans-serif;} h1, h2, h3, b {color:white;background-color:#525D76;} h1 {font-size:22px;} h2 {font-size:16px;} h3 {font-size:14px;} p {font-size:12px;} a {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 501 – Not Implemented</h1><hr class="line" /><p><b>Type</b> Status Report</p><p><b>Message</b> Not Implemented</p><p><b>Description</b> The server does not support the functionality required to fulfill the request.</p><hr class="line" /><h3>Apache Tomcat/9.0.98</h3></body></html>


In [15]:
# convert measurements units



## add gwp local


params = {
    'format': 'JSON'
}

res = requests.get("https://data.eco-platform.org/resource/flowproperties", headers=headers, params=params)
df = pd.DataFrame(res.json()["data"])

item = df.loc[df["uuid"] == "93a60a56-a3c8-11da-a746-0800200b9a66"]
item = {key: list(value.values())[0] for key, value in item.to_dict().items()}
print(item)
df

{'name': 'Mass', 'uuid': '93a60a56-a3c8-11da-a746-0800200b9a66', 'version': '03.00.000', 'classific': 'Technical flow properties', 'defUnitGrp': 'Masseneinheit', 'defUnit': 'kg', 'dsType': 'FlowProperty'}


name  \
0                                                Area   
1                                           Area*time   
2                           Carbon content (biogenic)   
3               Carbon content (biogenic) - packaging   
4                                            Duration   
5                   Energy ren. (net calorific value)   
6   Erneuerbare Primärenergie als Energieträger (P...   
7   Erneuerbare Primärenergie zur stofflichen Nutzung   
8                     Goods transport (mass*distance)   
9                               Gross calorific value   
10                                             Length   
11                                               Mass   
12                                Net calorific value   
13  Nicht erneuerbare Primärenergie zur stoffliche...   
14  Nicht-erneuerbare Primärenergie als Energieträ...   
15                                      Normal Volume   
16                                   Number of pieces   
17                                              Price   
18                                      Radioactivity   
19      Total nicht erneuerbare Primärenergie (PENRT)   
20                                             Volume   

                                    uuid    version  \
0   93a60a56-a3c8-19da-a746-0800200c9a66  03.00.000   
1   93a60a56-a3c8-21da-a746-0800200c9a66  03.00.000   
2   62e503ce-544a-4599-b2ad-bcea15a7bf20  03.00.003   
3   262a541b-209e-44cc-a426-33bce30de7b1  00.01.000   
4   c0447923-0e60-4b3c-97c2-a86dddd9eea5  03.00.000   
5   9b784a67-fbb1-4ad2-8774-30c39643b844  02.00.000   
6   f403679a-8f0c-4945-85ca-15c5d6d9cadf  25.00.000   
7   67cbf02a-2a4e-4668-941a-b4209c30b3c6  20.00.000   
8   838aaa20-0117-11db-92e3-0800200c9a66  03.00.000   
9   93a60a56-a3c8-14da-a746-0800200c9a66  03.00.000   
10  838aaa23-0117-11db-92e3-0800200c9a66  03.00.000   
11  93a60a56-a3c8-11da-a746-0800200b9a66  03.00.000   
12  93a60a56-a3c8-11da-a746-0800200c9a66  03.00.000   
13  59b3b1e8-813f-4647-81ca-8220ea2a4cdb  20.00.000   
14  15aa5df7-7905-4004-9b5f-a5a539c7aaae  25.00.000   
15  93a60a56-a3c8-13da-a746-0800200c9a66  03.00.000   
16  01846770-4cfe-4a25-8ad9-919d8d378345  03.00.000   
17  bcf81c0a-9fe9-4b88-a25b-99e0dd64ddaa  19.00.000   
18  93a60a56-a3c8-17da-a746-0800200c9a66  03.00.000   
19  70097bf8-cd55-4045-af5f-07144a8448ce  25.00.000   
20  93a60a56-a3c8-22da-a746-0800200c9a66  03.00.000   

                                            classific              defUnitGrp  \
0                           Technical flow properties          Fläche-Einheit   
1                           Technical flow properties      Units of area*time   
2                       Chemical composition of flows           Units of mass   
3                       Chemical composition of flows           Units of mass   
4                           Technical flow properties           Units of time   
5                           Technical flow properties         Units of energy   
6   Umweltliche Größen / EPD EN 15804 / LCI Indika...          Energieeinheit   
7   Umweltliche Größen / EPD EN 15804 / LCI Indika...         Units of energy   
8                           Technical flow properties    Units of mass*length   
9                           Technical flow properties         Units of energy   
10                          Technical flow properties         Units of length   
11                          Technical flow properties           Masseneinheit   
12                          Technical flow properties         Units of energy   
13  Umweltliche Größen / EPD EN 15804 / LCI Indika...         Units of energy   
14  Umweltliche Größen / EPD EN 15804 / LCI Indika...          Energieeinheit   
15                          Technical flow properties         Units of volume   
16                          Technical flow properties          Anzahl-Einheit   
17                           Economic flow properties                     NaN   
18                          Technical flow p

In [16]:
res = requests.get(f"https://data.eco-platform.org/resource/unitgroups/{uuid}", headers=headers, params=params)

print(res.text)


A unit group data set with the uuid e26d3411-373a-4cdd-af82-08db7e36184f cannot be found in the database


In [17]:
# uri = df.iloc[0]["uri"]
# params={}
# params['format'] = 'JSON'
# params['view'] = 'extended'


# print(uri)
# res = requests.get(uri, headers=headers, params=params)#, params=params)
#print(res.json().keys())
#print(json.dumps(res.json(), indent=4))

#pd.DataFrame(res.json())

exchanges shortDescription is inconsistent use uuid



publicationAndOwnership -> referenceToRegistrationAuthority as example

note the root node only shows an overview and dose not resolve the full data but only has the url and uuid where you can pull the full dataset

In [18]:

#res = requests.get(df.iloc[0]["uri"], headers=headers)#, params=params)
#print(df.iloc[0]["dataSources"][0]["uuid"])

# print(res.json()["LCIAResults"]["LCIAResult"].keys())
# pd.DataFrame(res.json()["LCIAResults"]["LCIAResult"][0])



## Amended  


### Python Setup 


You need to have [Python](https://www.python.org/) set up for this notebook to run. 

Just follow the instructions provided [here](https://www.python.org/downloads/). 

We recommend using a [virtual environment](https://docs.python.org/3/library/venv.html), but feel free to use whatever you like.


1. Create and activate a virtual environment:
   ```bash
   python3 -m venv py_env
   source py_env/bin/activate
   ```

2. Install the required packages:
   ```bash
   pip install ipykernel notebook numpy matplotlib pandas pyarrow ipympl
   ```

3. Add the virtual environment to Jupyter:
   ```bash
   python -m ipykernel install --user --name=py_env --display-name "Python (py_env)"
   ```

4. (Optional) Open the project in VSCode:
   ```bash
   code .
   ```

   - Ensure the virtual environment is active inside VSCode.
   - Select the kernel `Python (py_env)` when opening a notebook.

5. If using the integrated terminal in VSCode, activate the environment inside the terminal:
   ```bash
   source py_env/bin/activate
   ```

### Getting API Key

To get started, you'll need an access token. Follow these steps:

1. **Register for an API Key:**  
   - Go to the [registration page](https://data.eco-platform.org/registration.xhtml).
   - Complete the registration process if you haven’t already.

2. **Generate Your Token:**  
   - Scroll to the bottom of the page and click on **`My Profile`**.
   - Navigate to the **`API Key`** section.
   - Generate a new API key and copy it.

3. **Save the Token:**  
   - Create a `.env` file in your project folder.
   - Add the token to the file with the variable name `TOKEN`. The `.env` file should look something like this:  
     ```plaintext
     TOKEN=iY291bnRyeSI6IkR...aWFz
     ```